hub/bolt_transport.go

```diff
@@ -0,0 +1,239 @@
+package hub
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"sync"
+
+	bolt "go.etcd.io/bbolt"
+	"go.uber.org/atomic"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBoltBucketName = "updates"
+
+// BoltTransport implements the TransportInterface using the Bolt database
+type BoltTransport struct {
+	sync.RWMutex
+	db               *bolt.DB
+	bucketName       string
+	size             uint64
+	cleanupFrequency float64
+	pipes            map[*Pipe]struct{}
+	done             chan struct{}
+	lastSeq          atomic.Uint64
+}
+
+// NewBoltTransport create a new BoltTransport
+func NewBoltTransport(options *Options) (*BoltTransport, error) {
+	var err error
+	url := options.TransportURL
+	q := url.Query()
+	bucketName := defaultBoltBucketName
+	if q.Get("bucket_name") != "" {
+		bucketName = q.Get("bucket_name")
+	}
+
+	size := uint64(0)
+	if q.Get("size") != "" {
+		size, err = strconv.ParseUint(q.Get("size"), 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf(`invalid bolt "%s" dsn: parameter size: %w`, url, err)
+		}
+	}
+
+	cleanupFrequency := 0.3
+	if q.Get("cleanup_frequency") != "" {
+		cleanupFrequency, err = strconv.ParseFloat(q.Get("cleanup_frequency"), 64)
+		if err != nil {
+			return nil, fmt.Errorf(`invalid bolt "%s" dsn: parameter cleanup_frequency: %w`, url, err)
+		}
+	}
+
+	path := url.Path // absolute path (bolt:///path.db)
+	if path == "" {
+		path = url.Host // relative path (bolt://path.db)
+	}
+	if path == "" {
+		return nil, fmt.Errorf(`invalid bolt "%s" dsn: missing path`, url)
+	}
+
+	db, err := bolt.Open(path, 0600, nil)
+	if err != nil {
+		return nil, fmt.Errorf(`invalid bolt "%s" dsn: %w`, url, err)
+	}
+
+	return &BoltTransport{db: db, bucketName: bucketName, size: size, cleanupFrequency: cleanupFrequency, pipes: make(map[*Pipe]struct{}), done: make(chan struct{})}, nil
+}
+
+// Write pushes updates in the Transport
+func (t *BoltTransport) Write(update *Update) error {
+	select {
+	case <-t.done:
+		return ErrClosedTransport
+	default:
+	}
+
+	var (
+		err         error
+		closedPipes []*Pipe
+	)
+
+	t.RLock()
+
+	err = t.persist(update)
+
+	for pipe := range t.pipes {
+		if !pipe.Write(update) {
```

Comment:
should we write the update even if the persist failed?

Reply:
Good question. Both behaviors have benefits and drawbacks...

Reply:
I see 2 drawback:
- People consuming live streaming won't get the same data as people subscribing later.
- The Hub will return an error to the producer while update has been pushed to the subscribers....

Reply:
Indeed if it we send an error to the publisher, we must not dispatch the update. Let's change.